File: app/src/main/java/com/benmohammad/multithreading/demo/designthreadpool/ProducerConsumerBenchmarkUseCase.java
package com.benmohammad.multithreading.demo.designthreadpool;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.benmohammad.multithreading.common.BaseObservable;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public class ProducerConsumerBenchmarkUseCase extends BaseObservable<ProducerConsumerBenchmarkUseCase.Listener> {

    public static interface Listener {
        void onBenchmarkCompleted(Result result);
    }

    public static class Result {
        private final long mExecutionTime;
        private final int mNumOfReceivedMessages;

        public Result(long executionTime, int numOfReceivedMessages) {
            this.mExecutionTime = executionTime;
            this.mNumOfReceivedMessages = numOfReceivedMessages;
        }

        public long getExecutionTime() {
            return mExecutionTime;
        }

        public int getNumOfReceivedMessages() {
            return mNumOfReceivedMessages;
        }
    }

    public static final int NUM_OF_MESSAGES = 1000;
    public static final int BLOCKING_QUEUE_CAPACITY = 5;

    private final Object LOCK = new Object();
    private final Handler mUiHandler = new Handler(Looper.getMainLooper());
    private final AtomicInteger mNumOfThreads = new AtomicInteger(0);
    private final MyBlockingQueue mBlockingQueue = new MyBlockingQueue(BLOCKING_QUEUE_CAPACITY);

    private final ExecutorService mThreadPool = Executors.newCachedThreadPool(
            r -> {
                Log.d("ThreadFactory", "thread: " + mNumOfThreads.incrementAndGet());
                return new Thread(r);
            }
    );

    private int mNumOfFinishedThreads;
    private int mNumOfReceivedMessages;
    private long mStartTimestamp;

    public void startBenchmarkAndNotify() {
        synchronized (LOCK) {
            mNumOfReceivedMessages = 0;
            mNumOfFinishedThreads = 0;
            mStartTimestamp = System.currentTimeMillis();
            mNumOfThreads.set(0);
        }

        mThreadPool.execute(() -> {
            synchronized (LOCK) {
                while(mNumOfFinishedThreads < NUM_OF_MESSAGES) {
                    try {
                        LOCK.wait();
                    } catch (InterruptedException e)  {
                        return;
                    }
                }
            }

            notifySuccess();

        });

        mThreadPool.execute(() -> {
            for(int i = 0; i < NUM_OF_MESSAGES; i++) {
                startNewProducer(i);
            }
        });

        mThreadPool.execute(() -> {
            for(int i = 0; i < NUM_OF_MESSAGES; i++) {
                startNewConsumer();
            }
        });

    }


    private void startNewProducer(final int index) {
        mThreadPool.execute(() -> mBlockingQueue.put(index));
    }

    private void startNewConsumer() {
        mThreadPool.execute(() -> {
            int message = mBlockingQueue.take();
            synchronized (LOCK) {
                if(message != -1) {
                    mNumOfReceivedMessages++;
                }
                mNumOfFinishedThreads++;
                LOCK.notifyAll();
            }
        });
    }

    private void notifySuccess() {
        mUiHandler.post(() -> {
            Result result;
            synchronized (LOCK) {
                result = new Result(
                        System.currentTimeMillis() - mStartTimestamp,
                        mNumOfReceivedMessages
                );
                for(Listener listener : getListener()) {
                    listener.onBenchmarkCompleted(result);
                }
            }
        });
    }

}

File: app/src/main/java/com/benmohammad/multithreading/demo/designcoroutines/ProducerConsumerBenchmarkUseCase.kt
package com.benmohammad.multithreading.demo.designcoroutines

import android.os.Handler
import android.os.Looper
import com.benmohammad.multithreading.common.BaseObservable
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class ProducerConsumerBenchmarkUseCase : BaseObservable<ProducerConsumerBenchmarkUseCase.Listener>() {

    interface Listener {
        fun onBenchmarkCompleted(result: Result)
    }

    class Result(val executionTime : Long, val numOfreceivedMessages : Int)

    private val reentrantLock = ReentrantLock()
    private val lockCondition = reentrantLock.newCondition()

    private val uiHandler = Handler(Looper.getMainLooper())
    private val blockingQueue = MyBlockingQueue(BLOCKING_QUEUE_CAPACITY)
    private var numOfFinishedConsumers : Int = 0
    private var numOfReceivedMessages : Int = 0
    private var startTimestamp : Long = 0

    fun startBenchmarkAndNotify() {
        reentrantLock.withLock {
            numOfReceivedMessages = 0
            numOfFinishedConsumers = 0
            startTimestamp = System.currentTimeMillis()
        }


        Thread {
            reentrantLock.withLock {
                while(numOfFinishedConsumers < NUM_OF_MESSAGES) {
                    try {
                        lockCondition.await()
                    } catch (e: InterruptedException) {
                        return@Thread
                    }
                }
            }
         notifySuccess()

    }.start()

        Thread{
            for(i in 0 until NUM_OF_MESSAGES) {
                startNewProducer(i)
            }
        }.start()

        Thread{
            for(i in 0 until NUM_OF_MESSAGES) {
                startNewConsumer()
            }
        }.start()

    }


    fun startNewProducer(index: Int) {
      Thread{
          blockingQueue.put(index)
      }.start()
    }

    fun startNewConsumer() {
        Thread {
            val message = blockingQueue.take()
            reentrantLock.withLock {
                if(message !== -1) {
                    numOfReceivedMessages++
                }
                numOfFinishedConsumers++
                lockCondition.signalAll()
            }
        }.start()
    }

    private fun notifySuccess() {
        uiHandler.post {
            lateinit var result : Result
            reentrantLock.withLock {
                result = Result(
                        System.currentTimeMillis() - startTimestamp,
                        numOfReceivedMessages
                )
            }

            for(listener in listener) {
                listener.onBenchmarkCompleted(result)
            }



        }
    }



    companion object {
        private const val NUM_OF_MESSAGES = 1000
        private const val BLOCKING_QUEUE_CAPACITY = 5
    }
}